File: ios/QubitSDKModule/QubitSDKModule.m
//
//  QubitSDKModule.m
//  QubitSDKModule
//
//  Created by Michał Balawajder on 09/12/2019.
//  Copyright © 2019 Facebook. All rights reserved.
//

#import <Foundation/Foundation.h>
#import <React/RCTBridgeModule.h>

@interface RCT_EXTERN_REMAP_MODULE(QubitSDK, QubitSDKModule, NSObject)

RCT_EXTERN_METHOD(init:(NSString *)trackingId logLevel:(NSString *)logLevel)
RCT_EXTERN_METHOD(sendEvent:(NSString *)eventType eventBody:(NSDictionary *)eventBody)
RCT_EXTERN_METHOD(enableTracker:(BOOL)enable)
RCT_EXTERN_METHOD(getTrackingId:(RCTPromiseResolveBlock)resolver rejecter:(RCTPromiseRejectBlock)rejecter)
RCT_EXTERN_METHOD(getDeviceId:(RCTPromiseResolveBlock)resolver rejecter:(RCTPromiseRejectBlock)rejecter)
RCT_EXTERN_METHOD(getLookupData:(RCTPromiseResolveBlock)resolver rejecter:(RCTPromiseRejectBlock)rejecter)
RCT_EXTERN_METHOD(getExperiences:
                  (NSArray *) experienceIds
                  isVariationSet:(BOOL) isVariationSet
                  variation:(NSNumber *) variation
                  isPreviewSet:(BOOL) isPreviewSet
                  preview:(BOOL) preview
                  isIgnoreSegmentsSet:(BOOL) isIgnoreSegmentsSet
                  ignoreSegments:(BOOL) ignoreSegments
                  resolver:(RCTPromiseResolveBlock) resolver
                  rejecter:(RCTPromiseRejectBlock) rejecter)
RCT_EXTERN_METHOD(experienceShown:(NSString *) callback)
RCT_EXTERN_METHOD(getPlacement:
                  (NSString *) placementId
                  mode:(NSString *) mode
                  attributes:(NSString *) attributes
                  campaignId:(NSString *) campaignId
                  experienceId:(NSString *) experienceId
                  resolver:(RCTPromiseResolveBlock) resolver
                  rejecter:(RCTPromiseRejectBlock) rejecter)
RCT_EXTERN_METHOD(placementImpression:(NSString *) callback)
RCT_EXTERN_METHOD(placementClickthrough:(NSString *) callback)

@end

File: ios/QubitSDKModule/QubitSDKModule.swift
//
//  QubitSDKModule.swift
//  QubitSDKModule
//
//  Created by Michał Balawajder on 09/12/2019.
//

import Foundation
import QubitSDK

@objc(QubitSDKModule)
class QubitSDKModule: NSObject {
    
    @objc(init:logLevel:)
    func `init`(trackingId: String, logLevel: String) {
        QubitSDK.start(withTrackingId: trackingId, logLevel: QBLogLevel.logLevel(from: logLevel), queuePriority: .background)
    }
    
    @objc(sendEvent:eventBody:)
    func sendEvent(eventType: String, eventBody: [String: Any]) {
        QubitSDK.sendEvent(type: eventType, dictionary: eventBody)
    }
    
    @objc(enableTracker:)
    func enableTracker(enable: Bool) {
        QubitSDK.enableTracker(enable: enable)
    }
    
    @objc(getTrackingId:rejecter:)
    func getTrackingId(resolver: RCTPromiseResolveBlock, rejecter: RCTPromiseRejectBlock) {
        resolver(QubitSDK.trackingId)
    }
    
    @objc(getDeviceId:rejecter:)
    func getDeviceId(resolver: RCTPromiseResolveBlock, rejecter: RCTPromiseRejectBlock) {
        resolver(QubitSDK.deviceId)
    }
    
    @objc(getLookupData:rejecter:)
    func getLookupData(resolver: RCTPromiseResolveBlock, rejecter: RCTPromiseRejectBlock) {
        guard let lookup = QubitSDK.getLookupData() else {
            rejecter("Error", "QubitModuleSDK: getLookupData returned nil; probably no lookup data yet", nil)
            return
        }
        resolver(lookup.dictionary)
    }
    
    @objc(getExperiences:isVariationSet:variation:isPreviewSet:preview:isIgnoreSegmentsSet:ignoreSegments:resolver:rejecter:)
    func getExperiences(experienceIds: [Int], isVariationSet: Bool, variation: Int, isPreviewSet: Bool, preview: Bool, isIgnoreSegmentsSet: Bool, ignoreSegments: Bool,  resolver: @escaping RCTPromiseResolveBlock, rejecter: @escaping RCTPromiseRejectBlock) {
        QubitSDK.fetchExperiences(withIds: experienceIds, onSuccess: { result in
            resolver(result.map({ $0.asDictionary }))
        }, onError: { error in
            rejecter("Error", "QubitSDKModule: getExperiences failed.", error)
        }, preview: isPreviewSet ? preview : false,
           ignoreSegments: isIgnoreSegmentsSet ? ignoreSegments : false,
           variation: isVariationSet ? NSNumber(integerLiteral: variation) : nil)
    }
    
    @objc(experienceShown:)
    func experienceShown(callback: String) {
        QBExperienceEntityCallback(callback: callback).shown()
    }

    @objc(getPlacement:mode:attributes:campaignId:experienceId:resolver:rejecter:)
    func getPlacement(placementId: String, mode: String, attributes: String, campaignId: String, experienceId: String, resolver: @escaping RCTPromiseResolveBlock, rejecter: @escaping RCTPromiseRejectBlock) {
        QubitSDK.getPlacement(with: placementId,
                              mode: mode,
                              attributes: attributes,
                              campaignId: campaignId,
                              experienceId: experienceId,
                              onSuccess: { result in
                                resolver( result.asDictionary )
                              }, onError: { error in
                                rejecter("Error", "QubitSDKModule: getPlacement failed.", error)
                              })
    }

    @objc(placementImpression:)
    func placementImpression(callback: String) {
        QBPlacementEntityCallback(impressionUrl: callback).impression()
    }

    @objc(placementClickthrough:)
    func placementClickthrough(callback: String) {
        QBPlacementEntityCallback(clickthroughUrl: callback).clickthrough()
    }
    
    @objc
    static func requiresMainQueueSetup() -> Bool {
        return true
    }
}
